jupyter_server/tests/test_firewall.py

```diff
@@ -0,0 +1,318 @@
+    "info": {"title": "Test specs", "version": "0.0.1"},
+    "paths": {
+        "/pet": {
+            "put": {
+                "requestBody": {
+                    "content": {
+                        "application/json": {
+                            "schema": {"$ref": "#/components/schemas/Pet"}
+                        },
+                    },
+                    "required": True,
+                },
+                "responses": {"400": {"description": ""}},
+            }
+        },
+        "/pet/findByTags": {
+            "get": {
+                "parameters": [
+                    {
+                        "name": "tags",
+                        "in": "query",
+                        "description": "Tags to filter by",
+                        "required": True,
+                        "style": "form",
+                        "schema": {"type": "array", "items": {"type": "string"}},
+                    }
+                ],
+                "responses": {
+                    "200": {
+                        "description": "successful operation",
+                    }
+                },
+            }
+        },
+        "/pet/{petId}": {
+            "get": {
+                "parameters": [
+                    {
+                        "name": "petId",
+                        "in": "path",
+                        "description": "ID of pet to return",
+                        "required": True,
+                        "schema": {"type": "integer", "format": "int64"},
+                    }
+                ],
+                "responses": {"200": {"description": ""}},
+            },
+        },
+    },
+    "components": {
+        "schemas": {
+            "Pet": {
+                "required": ["name"],
+                "type": "object",
+                "properties": {
+                    "id": {"type": "integer", "format": "int64"},
+                    "name": {"type": "string", "example": "doggie"},
+                },
+            },
+        }
+    },
+}
+
+spec2 = {
+    "openapi": "3.0.1",
+    "info": {"title": "Test specs", "version": "0.0.1"},
+    "paths": {
+        "/pet": {
+            "put": {
+                "responses": {"400": {"description": ""}},
+            }
+        },
+        "/user/{anyId}": {
+            "get": {
+                "parameters": [
+                    {
+                        "name": "anyId",
+                        "in": "path",
+                        "required": True,
+                        "schema": {"type": "string"},
+                    }
+                ],
+                "responses": {"200": {"description": ""}},
+            }
+        },
+    },
+}
+
+
+def format_id(val):
+    if isinstance(val, HTTPServerRequest):
+        return f"{val.method}-{val.uri}-{val.body.decode('utf-8') or None}"
+
+
+@pytest.mark.parametrize(
+    "base_url, server_request, expected",
+    (
+        (
+            "/",
+            HTTPServerRequest(
```

Comment:
i typically try to avoid classes inside parametrized stuff: the tracebacks are pretty hard to follow with `HTTPRequest-1`, etc.